driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +184,356 @@ public int getMaxWireVersion() {
+            }
+            if (doClose) {
+                doClose();
+            }
+        }
+
+        /**
+         * This method is never executed concurrently with either itself or other operations
+         * demarcated by {@link #tryStartOperation()}/{@link #endOperation()}.
+         */
+        private void doClose() {
+            if (skipReleasingServerResourcesOnClose) {
+                serverCursor = null;
             }
-            endOperationInProgress();
 
-            if (result.getResults().isEmpty()) {
-                callback.onResult(null, null);
+            if (getServerCursor() != null) {
+                executeWithConnection(((connection, t) -> {
+                    assertTrue(connection != null || t != null);
+                    if (t != null) {
+                        // guarantee that regardless of exceptions, `serverCursor` is null and client resources are released
+                        serverCursor = null;
+                        releaseClientResources();
+                    } else {
+                        releaseServerAndClientResources(connection);
+                        releaseServerResources(connection);
+                        connection.release();
+                    }
+                }));
             } else {
-                callback.onResult(result.getResults(), null);
+                releaseClientResources();
             }
         }
-    }
 
-    private CommandCursorResult<T> initFromCommandCursorDocument(
-            final ServerAddress serverAddress,
-            final String fieldNameContainingBatch,
-            final BsonDocument commandCursorDocument) {
-        CommandCursorResult<T> cursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch, commandCursorDocument);
-        cursor.set(cursorResult.getServerCursor());
-        count.addAndGet(cursorResult.getResults().size());
-        postBatchResumeToken = cursorResult.getPostBatchResumeToken();
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", cursorResult.getResults().size(),
-                    cursorResult.getCursorId(), cursorResult.getServerAddress()));
+        void checkCorruptedConnection() {
+            // if `pinnedConnection` is corrupted, then we cannot kill `serverCursor` via such a connection
+            AsyncConnection localPinnedConnection = pinnedConnection;
+            if (localPinnedConnection != null) {
+                skipReleasingServerResourcesOnClose = true;
+            }
         }
-        return cursorResult;
-    }
 
-    private void logQueryResult(final CommandCursorResult<T> result) {
-        LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", result.getResults().size(),
-                result.getCursorId(), result.getServerAddress()));
-    }
+        void executeWithConnection(final AsyncCallableWithConnection callable) {
```

Comment:
In general, it's easier to understand reference counting when the code that releases the reference is as close as possible to the code that retains it.  But here, this method retains the connection, while `callable` is responsible for releasing it.  Let's change it so that this method is responsible for releasing.

I wasn't totally clear on how to do that though, so I ended up with a [patch](https://github.com/jyemin/mongo-java-driver/commits/JAVA-5159-4-refactor) trying to figure it out (the patch is not complete, as there are now two `executeWithConnection` methods. It's just to express the idea concretely.) Please have a look.

Reply:
Added to #401